### PR TITLE
MGMT-4314 KubeAPI: InstallEnv support for nmstate config

### PR DESCRIPTION
**Pull request by nmagnezi**
Adding NMState config support in KubeAPI.
To link between an InstallEnv to NMState (either one or more):
1. **InstallEnv CR:** add a label to nmStateConfigLabelSelector with a user defined name and value.
2. **NMState CR:** Specify the same label + value in Object metadata.

Upon InstallEnv creation, the InstallEnv controller will search by label+value for matching NMState resources and construct a config to be sent as StaticNetworkConfig as a part of ImageCreateParams. The backend does all validations, and currently, there is no handling of configuration conflicts (e.g., two nmstate resources using the same MAC address).

The InstallEnv controller will watch for NMState config creation/changes and search for corresponding InstallEnv resources to reconcile since we need to regenerate the image for those.

NMState CRD handles the config with custom marshaling/unmarshaling, ported from:
https://github.com/nmstate/kubernetes-nmstate/blob/master/api/shared/nodenetworkstate_util.go

Examples may be found here:
https://github.com/openshift/assisted-service/blob/master/docs/crds/installEnv.yaml
https://github.com/openshift/assisted-service/blob/master/docs/crds/nmstate.yaml

### CODE REVIEW

**Comment by nmagnezi** [comment]
/assign @yevgeny-shnaidman
please take a look and let me know if you have any remarks.

File: internal/controller/controllers/installenv_controller.go
Comment (filanov) [inline internal/controller/controllers/installenv_controller.go:153]: why is that a const? we can probably have multiple labels here no?

File: internal/controller/controllers/installenv_controller.go
Comment (filanov) [inline internal/controller/controllers/installenv_controller.go:163]: suggesting to use make it will be faster, you know the size by `len(nmStateConfig.Spec.MacInterfaceMap)`

File: internal/controller/controllers/installenv_controller_test.go
Comment (filanov) [inline internal/controller/controllers/installenv_controller_test.go:225]: if you have the same definition for all the related tests then you can put them under a single context and define the variables there.

File: internal/controller/controllers/installenv_controller.go
Comment (nmagnezi) [reply internal/controller/controllers/installenv_controller.go:153]: Good question.
@mhrivnak, @avishayt,
Currently, this adds a label selector in InstallEnv spec and a label in NMStateConfig spec.
Do we expect a single InstallEnv to match with multiple NMstate labels?
alternately, do we expect a single NMStateConfig to match multiple InstallEnvs?

File: internal/controller/controllers/common.go
Comment (nmagnezi) [reply internal/controller/controllers/common.go]: Done.
Note that we'll still need to change it manually for the `+kubebuilder` scaffold lines

File: internal/controller/controllers/installenv_controller.go
Comment (nmagnezi) [reply internal/controller/controllers/installenv_controller.go]: Done

File: internal/controller/controllers/installenv_controller_test.go
Comment (nmagnezi) [reply internal/controller/controllers/installenv_controller_test.go]: Done

File: internal/controller/controllers/installenv_controller.go
Comment (mhrivnak) [reply internal/controller/controllers/installenv_controller.go:153]: Just at a high level, we still don't have consensus on the details of how to use nmstate in this API, so just be prepared for some flexibility.

> Do we expect a single InstallEnv to match with multiple NMstate labels?

Yes. As in static networking where IPs are assigned. Or different classes of hardware in the same environment require different configs.

> alternately, do we expect a single NMStateConfig to match multiple InstallEnvs?

No. Each environment defines its own configuration.

Comment (nmagnezi) [reply internal/controller/controllers/installenv_controller.go]: Done.
both label name and value are now user-configurable.
CR examples updated to reflect that.

**Review by filanov — APPROVED** [review]

**Comment by filanov** [comment]
@osherdp is there an automatic e2e test?

**Comment by nmagnezi** [comment]
/unhold
[
Tests passed](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_assisted-service/1316/pull-ci-openshift-assisted-service-master-e2e-metal-assisted-kube-api/1377258783100637184)

**Comment by osherdp** [comment]
> @osherdp: I saw an ad in a shop window, "Television for sale, $1, volume stuck on full", I thought, "I can't turn that down".

It's like they have a group of dads writing all those jokes for them

File: internal/controller/api/v1alpha1/nmstate_config_types.go
Comment (mhrivnak) [inline internal/controller/api/v1alpha1/nmstate_config_types.go:32]: We should expand on exactly how this will be used. Something like:

```
// Interfaces is an array of interface objects containing the name and MAC 
// address for interfaces that are referenced in the raw nmstate config YAML.
// Interfaces listed here will be automatically renamed in the nmstate config
// YAML to match the real device name that is observed to have the
// corresponding MAC address. At least one interface must be listed so that it
// can be used to identify the correct host, which is done by matching any MAC
// address in this list to any MAC address observed on the host.
```

File: internal/controller/api/v1alpha1/nmstate_config_types.go
Comment (mhrivnak) [inline internal/controller/api/v1alpha1/nmstate_config_types.go:27]: Should use a [validation marker](https://book.kubebuilder.io/reference/markers/crd-validation.html#crd-validation) also to enforce this
```
+kubebuilder:validation:Pattern
```

File: internal/controller/api/v1alpha1/nmstate_config_types.go
Comment (mhrivnak) [inline internal/controller/api/v1alpha1/nmstate_config_types.go:35]: We want the experience of expressing nmstate config to be similar to that of the `NodeNetworkConfigurationPolicy`, where the nmstate yaml is basically in-lined as part of the Spec. See the [example in their README](https://github.com/nmstate/kubernetes-nmstate/blob/master/README.md).

You can see some details on how they're accomplishing that in [their code](https://github.com/nmstate/kubernetes-nmstate/blob/master/api/shared/state_types.go). Making this field a `string` type would be a different experience.

Personally I prefer the simpler field name "config". Forcing the content to be in-line will make it implicitly YAML, so a more specific noun is warranted. And I think "network" is implied since the resource is called `NMStateConfig`. But I'm open to other opinions. Naming is hard.

File: internal/controller/api/v1alpha1/nmstate_config_types.go
Comment (mhrivnak) [reply internal/controller/api/v1alpha1/nmstate_config_types.go:32]: We can also use a validation marker to enforce a minimum of 1 item in the array.

**Review by mhrivnak — CHANGES_REQUESTED** [review]
Nir, LMK if it would help to chat through any of this or what feedback you have.

@ronniel1 @avishayt @celebdor please make sure this aligns with your understanding of what we discussed.

File: internal/controller/api/v1alpha1/nmstate_config_types.go
Comment (nmagnezi) [reply internal/controller/api/v1alpha1/nmstate_config_types.go]: Done

File: internal/controller/api/v1alpha1/nmstate_config_types.go
Comment (nmagnezi) [reply internal/controller/api/v1alpha1/nmstate_config_types.go]: Done

File: internal/controller/api/v1alpha1/nmstate_config_types.go
Comment (nmagnezi) [reply internal/controller/api/v1alpha1/nmstate_config_types.go]: Done

**Review by filanov — APPROVED** [review]

**Comment by nmagnezi** [comment]
/hold 
fedora copr is back. checking something locally

File: internal/controller/api/v1alpha1/nmstate_config_types.go
Comment (filanov) [inline internal/controller/api/v1alpha1/nmstate_config_types.go:49]: you are missing `	// +kubebuilder:validation:XPreserveUnknownFields`

File: internal/controller/api/v1alpha1/nmstate_config_types.go
Comment (nmagnezi) [reply internal/controller/api/v1alpha1/nmstate_config_types.go]: Done

File: internal/controller/api/v1alpha1/nmstate_config_types.go
Comment (filanov) [inline internal/controller/api/v1alpha1/nmstate_config_types.go:40]: Please add a comment explaining that we have a specific implementation for the unmarshal

**Comment by nmagnezi** [comment]
tests are expected to fail, until we merge https://github.com/openshift/assisted-service/pull/1455

File: internal/controller/api/v1alpha1/nmstate_config_types.go
Comment (nmagnezi) [reply internal/controller/api/v1alpha1/nmstate_config_types.go:40]: done

**Comment by nmagnezi** [comment]
rebased to include a CI fix https://github.com/openshift/assisted-service/pull/1466

**Comment by nmagnezi** [comment]
regenerated and rebased to resolve a conflict in `role.yaml`

**Review by filanov — APPROVED** [review]

**Comment by filanov** [comment]
@nmagnezi another subsystem test failed please take a look

**Comment by filanov** [comment]
```
INFO[2021-04-08T06:08:38Z] • Failure [151.507 seconds]     
INFO[2021-04-08T06:08:38Z] [kube-api]cluster installation               
INFO[2021-04-08T06:08:38Z] /go/src/github.com/openshift/assisted-service/subsystem/kubeapi_test.go:305 
INFO[2021-04-08T06:08:38Z]   deploy clusterDeployment with agents and wait for ready [It] 
INFO[2021-04-08T06:08:38Z]   /go/src/github.com/openshift/assisted-service/subsystem/kubeapi_test.go:319 
INFO[2021-04-08T06:08:38Z]                                              
skipped 137 lines unfold_more
INFO[2021-04-08T06:08:38Z]                                              
INFO[2021-04-08T06:08:38Z] [Fail] [kube-api]cluster installation [It] deploy clusterDeployment with agents and wait for ready  
INFO[2021-04-08T06:08:38Z] /go/src/github.com/openshift/assisted-service/subsystem/kubeapi_test.go:353 
INFO[2021-04-08T06:08:38Z]                                              
INFO[2021-04-08T06:08:38Z] Ran 12 of 174 Specs in 310.353 seconds 
INFO[2021-04-08T06:08:38Z] FAIL! -- 11 Passed | 1 Failed | 0 Pending | 162 Skipped 
INFO[2021-04-08T06:08:38Z] --- FAIL: TestSubsystem (310.96s)            
INFO[2021-04-08T06:08:38Z] FAIL                                         
INFO[2021-04-08T06:08:38Z] FAIL	github.com/openshift/assisted-service/subsystem	314.400s 
INFO[2021-04-08T06:08:38Z] FAIL                                      
```

**Comment by nmagnezi** [comment]
> ```
> INFO[2021-04-08T06:08:38Z] • Failure [151.507 seconds]     
> INFO[2021-04-08T06:08:38Z] [kube-api]cluster installation               
> INFO[2021-04-08T06:08:38Z] /go/src/github.com/openshift/assisted-service/subsystem/kubeapi_test.go:305 
> INFO[2021-04-08T06:08:38Z]   deploy clusterDeployment with agents and wait for ready [It] 
> INFO[2021-04-08T06:08:38Z]   /go/src/github.com/openshift/assisted-service/subsystem/kubeapi_test.go:319 
> INFO[2021-04-08T06:08:38Z]                                              
> skipped 137 lines unfold_more
> INFO[2021-04-08T06:08:38Z]                                              
> INFO[2021-04-08T06:08:38Z] [Fail] [kube-api]cluster installation [It] deploy clusterDeployment with agents and wait for ready  
> INFO[2021-04-08T06:08:38Z] /go/src/github.com/openshift/assisted-service/subsystem/kubeapi_test.go:353 
> INFO[2021-04-08T06:08:38Z]                                              
> INFO[2021-04-08T06:08:38Z] Ran 12 of 174 Specs in 310.353 seconds 
> INFO[2021-04-08T06:08:38Z] FAIL! -- 11 Passed | 1 Failed | 0 Pending | 162 Skipped 
> INFO[2021-04-08T06:08:38Z] --- FAIL: TestSubsystem (310.96s)            
> INFO[2021-04-08T06:08:38Z] FAIL                                         
> INFO[2021-04-08T06:08:38Z] FAIL	github.com/openshift/assisted-service/subsystem	314.400s 
> INFO[2021-04-08T06:08:38Z] FAIL                                      
> ```

to be handled by https://github.com/openshift/assisted-service/pull/1472

**Pull request merged**
